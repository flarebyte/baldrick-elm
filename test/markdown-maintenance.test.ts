import {
  getCommandHelp,
  getMakefileCommands,
  getZshAliases,
  maintenanceMd,
} from '../src/markdown-maintenance.js';
import { libCoreProject } from './fixture-core-project.js';

describe('Maintenance documentation', () => {
  it('normalizes MAINTENANCE.md', () => {
    expect(maintenanceMd(libCoreProject)).toMatchSnapshot();
  });

  it('produces a list of npm scripts', () => {
    expect(getMakefileCommands(libCoreProject)).toMatchSnapshot();
  });

  it('produces a list of zsh aliases', () => {
    expect(getZshAliases()).toMatchInlineSnapshot(
      `"alias gcf='git add . && git commit -F .message && rm .message'"`
    );
  });

  it('produces an help for the commands', () => {
    expect(getCommandHelp(libCoreProject)).toMatchInlineSnapshot(`
      "Commands:
      act                  Run GitHub Actions inside a docker container
      gcf                  Git commit a message that has been saved in the .message file
      make analyze         Find problems in Elm code
      make build           Transpile all the Elm source code to javascript
      make doc             Generate the markdown documentation for the Elm project
      make github          Enable useful features for the github project repository
      make help            Summarize all the yarn and shell commands
      make install-global  Install some dependencies globally
      make md              Checks that the markdown documents follows some consistent guidelines
      make md:fix          Modify the markdown documents to ensure they follow some consistent guidelines
      make norm            Normalize the code structure using baldrick (global version)
      make preview-doc     Generate the markdown documentation for the Elm project
      make ready           Run a sequence of commands to check that the library is ready to be published
      make reset           Delete the dist and report folder
      make test            Run the unit tests"
    `);
  });
});
