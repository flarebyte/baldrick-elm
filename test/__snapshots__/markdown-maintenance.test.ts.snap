// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maintenance documentation normalizes MAINTENANCE.md 1`] = `
"# Maintenance of the code

## Overall workflow

The typical developer workflow goes as follow:

| Mode                   | Code analysis | Testing                          | Building   | Publishing |
| ---------------------- | ------------- | -------------------------------- | ---------- | ---------- |
| Checking               | make analyze  | make test                        | make build | make ready |
| Fixing                 | Fix the code  | Update dependencies and make doc |            |            |
| Continuous integration | make build    |                                  |            |            |

## Commands

### Run GitHub Actions locally

> Run GitHub Actions inside a docker container

__Motivation:__ Test GitHub Actions locally

__When to use it:__ When changing github actions

__Run:__ \`\`\`act\`\`\`

__From package:__ [act](https://github.com/nektos/act) of [brew](https://docs.brew.sh/) :  Run GitHub Actions locally

---

### Build the library

> Transpile all the Elm source code to javascript

__Motivation:__ The code has to be build before been published

__When to use it:__ Before building

__Run:__ \`\`\`make build\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Generate the documentation

> Generate the markdown documentation for the Elm project

__Motivation:__ Good documentation is essential for developer experience

__When to use it:__ Before publishing

__Run:__ \`\`\`make doc\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Preview the documentation

> Generate the markdown documentation for the Elm project

__Motivation:__ Good documentation is essential for developer experience

__When to use it:__ Before publishing

__Run:__ \`\`\`make preview-doc\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Update github repository

> Enable useful features for the github project repository

__Motivation:__ Create consistent settings

__When to use it:__ After creating

__Run:__ \`\`\`make github\`\`\`

__From package:__ [gh](https://cli.github.com/) of [brew](https://docs.brew.sh/) :  GitHub CLI brings GitHub to your terminal

---

### Static code analysis

> Find problems in Elm code

__Motivation:__ Make the code more consistent and avoid bugs

__When to use it:__ Before compilation

__Run:__ \`\`\`make analyze\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Beautify Elm source code

> Formats Elm source code according to a standard set of rules based on the official Elm Style Guide

__Motivation:__ Make the code more consistent and avoid bugs

__When to use it:__ Before compilation

__Run:__ \`\`\`make beautify\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Detects Elm code API changes

> See what changed in the package between versions

__Motivation:__ Sometimes a MAJOR change is not actually very big, so this can help you plan your upgrade timelines

__When to use it:__ Before compilation

__Run:__ \`\`\`make diff\`\`\`

__From package:__ [elm](https://guide.elm-lang.org/install/elm.html) of [npm](https://www.npmjs.com/) :  Dependency management

---

### Markdown check

> Checks that the markdown documents follows some consistent guidelines

__Motivation:__ Make the markdown documents consistent in style

__When to use it:__ Before publishing

__Run:__ \`\`\`make md\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Markdown fix

> Modify the markdown documents to ensure they follow some consistent guidelines

__Motivation:__ Make the markdown documents consistent in style

__When to use it:__ Before publishing

__Run:__ \`\`\`make md-fix\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Install local dependencies

> Install some dependencies

__Motivation:__ Before running most commands

__When to use it:__ Before building

__Run:__ \`\`\`make install\`\`\`

__From package:__ [elm](https://guide.elm-lang.org/install/elm.html) of [npm](https://www.npmjs.com/) :  Dependency management

---

### Install global dependencies

> Install some dependencies globally

__Motivation:__ Before running most commands

__When to use it:__ Before building

__Run:__ \`\`\`make install-global\`\`\`

__From package:__ [yarn](https://classic.yarnpkg.com/en/) of [npm](https://www.npmjs.com/) :  Dependency management

---

### Ready for publishing

> Run a sequence of commands to check that the library is ready to be published

__Motivation:__ Detect quality flaws before pushing the code

__When to use it:__ Before pushing a branch

__Run:__ \`\`\`make ready\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Reset distribution and report folders

> Delete the dist and report folder

__Motivation:__ Start from a clean slate

__When to use it:__ Before building

__Run:__ \`\`\`make reset\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Unit testing

> Run the unit tests

__Motivation:__ Check that the units of code behave as intended

__When to use it:__ After compilation, before build

__Run:__ \`\`\`make test\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Help for commands

> Summarize all the yarn and shell commands

__Motivation:__ Assist the developer in quickly finding commands

__When to use it:__ Before running a command

__Run:__ \`\`\`make help\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Git commit from file

> Git commit a message that has been saved in the .message file

__Motivation:__ Quicker commit for pre-defined use cases

__When to use it:__ When commit to github

__Run:__ \`\`\`gcf\`\`\`

__From package:__ [zsh](https://www.zsh.org/) of [brew](https://docs.brew.sh/) :  Shell designed for interactive use

---

### Reset generated folders

> Delete the generated folder

__Motivation:__ Start generation from a clean slate

__When to use it:__ Before generation

__Run:__ \`\`\`make reset-generated\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Prepare scripts for code generation

> Generate the scripts used for code generation

__Motivation:__ The generation scripts contain a fair amount of boilerplate code that can be easily generated

__When to use it:__ Before generation

__Run:__ \`\`\`make pre-generate\`\`\`

__From package:__ [zsh](https://www.zsh.org/) of [brew](https://docs.brew.sh/) :  Shell designed for interactive use

---

### Generate some Elm Code

> Generate some of the boilerplate code for the Elm project

__Motivation:__ Standardize and boost the development process

__When to use it:__ When changing model

__Run:__ \`\`\`make generate\`\`\`

__From package:__ [zsh](https://www.zsh.org/) of [brew](https://docs.brew.sh/) :  Shell designed for interactive use

---

### Generate some code in the console

> Generate some of the boilerplate code that has to be manually added

__Motivation:__ Boost the development process with contextual snippets

__When to use it:__ When changing model

__Run:__ \`\`\`make assist\`\`\`

__From package:__ [zsh](https://www.zsh.org/) of [brew](https://docs.brew.sh/) :  Shell designed for interactive use

---

### Normalize the project with baldrick-whisker

> Generate some scaffolding using some handlebars templates

__Motivation:__ Externalize scaffolding outside close source code

__When to use it:__ Before running normalization

__Run:__ \`\`\`make whisker-norm\`\`\`

__From package:__ [makefile](https://opensource.com/article/18/8/what-how-makefile) of [brew](https://docs.brew.sh/) :  Build management

---

### Normalize the code structure

> Normalize the code structure using baldrick (global version)

__Motivation:__ Create a consistent developer experience

__When to use it:__ When changing github actions

__Run:__ \`\`\`make norm\`\`\`

__From package:__ [baldrick-elm](https://github.com/flarebyte/baldrick-elm) of [npm](https://www.npmjs.com/) :  Elm scaffolding assistant

---"
`;

exports[`Maintenance documentation produces a list of npm scripts 1`] = `
Array [
  Object {
    "makeLines": Array [],
    "name": "build",
    "parentMakeTask": "test beautify doc",
    "title": "Build the library",
  },
  Object {
    "makeLines": Array [
      "elm make --docs=documentation.json",
    ],
    "name": "doc",
    "parentMakeTask": undefined,
    "title": "Generate the documentation",
  },
  Object {
    "makeLines": Array [
      "elm-doc-preview",
    ],
    "name": "preview-doc",
    "parentMakeTask": undefined,
    "title": "Preview the documentation",
  },
  Object {
    "makeLines": Array [
      "gh repo edit --delete-branch-on-merge --enable-squash-merge",
    ],
    "name": "github",
    "parentMakeTask": undefined,
    "title": "Update github repository",
  },
  Object {
    "makeLines": Array [
      "elm-analyse -s -o",
    ],
    "name": "analyze",
    "parentMakeTask": undefined,
    "title": "Static code analysis",
  },
  Object {
    "makeLines": Array [
      "elm-format src/ --yes",
      "elm-format tests/ --yes",
    ],
    "name": "beautify",
    "parentMakeTask": undefined,
    "title": "Beautify Elm source code",
  },
  Object {
    "makeLines": Array [
      "elm diff",
    ],
    "name": "diff",
    "parentMakeTask": undefined,
    "title": "Detects Elm code API changes",
  },
  Object {
    "makeLines": Array [
      "npx baldrick-dev-ts@latest markdown check",
      "npx baldrick-dev-ts@latest markdown check -s .github/",
    ],
    "name": "md",
    "parentMakeTask": undefined,
    "title": "Markdown check",
  },
  Object {
    "makeLines": Array [
      "npx baldrick-dev-ts@latest markdown fix",
      "npx baldrick-dev-ts@latest markdown fix -s .github/",
    ],
    "name": "md-fix",
    "parentMakeTask": undefined,
    "title": "Markdown fix",
  },
  Object {
    "makeLines": Array [
      "elm make",
      "pushd tests && elm make && popd",
    ],
    "name": "install",
    "parentMakeTask": undefined,
    "title": "Install local dependencies",
  },
  Object {
    "makeLines": Array [
      "yarn global add elm-format@0.8.4",
      "yarn global add elm-review",
      "yarn global add elm-upgrade",
      "yarn global add elm-doc-preview",
      "yarn global add elm-analyse",
      "yarn global add elm-test",
    ],
    "name": "install-global",
    "parentMakeTask": undefined,
    "title": "Install global dependencies",
  },
  Object {
    "makeLines": Array [],
    "name": "ready",
    "parentMakeTask": "analyze test",
    "title": "Ready for publishing",
  },
  Object {
    "makeLines": Array [
      "rm -rf elm-stuff",
      "rm -rf tests/elm-stuff",
    ],
    "name": "reset",
    "parentMakeTask": undefined,
    "title": "Reset distribution and report folders",
  },
  Object {
    "makeLines": Array [
      "elm-test",
    ],
    "name": "test",
    "parentMakeTask": undefined,
    "title": "Unit testing",
  },
  Object {
    "makeLines": Array [
      "cat commands.txt",
    ],
    "name": "help",
    "parentMakeTask": undefined,
    "title": "Help for commands",
  },
  Object {
    "makeLines": Array [
      "rm -rf generated",
    ],
    "name": "reset-generated",
    "parentMakeTask": undefined,
    "title": "Reset generated folders",
  },
  Object {
    "makeLines": Array [
      "npx baldrick-whisker@latest render script/data/project.json script/template/generate.hbs script/generate.sh",
      "npx baldrick-whisker@latest render script/data/project.json script/template/assist.hbs script/assist.sh",
    ],
    "name": "pre-generate",
    "parentMakeTask": undefined,
    "title": "Prepare scripts for code generation",
  },
  Object {
    "makeLines": Array [
      "mkdir generated",
      "sh script/generate.sh",
      "make beautify",
      "make test",
    ],
    "name": "generate",
    "parentMakeTask": "reset-generated pre-generate",
    "title": "Generate some Elm Code",
  },
  Object {
    "makeLines": Array [
      "sh script/assist.sh",
    ],
    "name": "assist",
    "parentMakeTask": "pre-generate",
    "title": "Generate some code in the console",
  },
  Object {
    "makeLines": Array [
      "mkdir -p .vscode",
      "mkdir -p .github/workflows",
      "mkdir -p .github/ISSUE_TEMPLATE",
      "mkdir -p src",
      "mkdir -p tests",
      "mkdir -p script",
      "mkdir -p script/data",
      "mkdir -p script/schema",
      "mkdir -p script/template",
      "test -f \\"elm.json\\" || npx baldrick-whisker@latest object elm.json github:flarebyte:baldrick-reserve:data/elm/src-elm.json",
      "test -f \\".vscode/settings.json\\" || npx baldrick-whisker@latest object .vscode/settings.json github:flarebyte:baldrick-reserve:data/elm/vscode-settings.json",
      "test -f \\"script/data/project.json\\" || npx baldrick-whisker@latest object script/data/project.json github:flarebyte:baldrick-reserve:data/elm/project.json",
      "npx baldrick-whisker@latest object --no-ext .vscode/baldrick.code-snippets.json github:flarebyte:baldrick-reserve:data/elm/snippet.yaml",
      "npx baldrick-whisker@latest object .github/ISSUE_TEMPLATE/bug_report.yaml github:flarebyte:baldrick-reserve:data/elm/bug-report.yaml",
      "npx baldrick-whisker@latest object .github/ISSUE_TEMPLATE/feature_request.yaml github:flarebyte:baldrick-reserve:data/elm/feature-request.yaml",
      "npx baldrick-whisker@latest object .github/workflows/main.yml github:flarebyte:baldrick-reserve:data/elm/workflow-main.yml",
      "npx baldrick-whisker@latest object script/schema/project.schema.json github:flarebyte:baldrick-reserve:data/elm/project.schema.json",
      "npx baldrick-whisker@latest render elm.json github:flarebyte:baldrick-reserve:template/code-of-conduct.hbs CODE_OF_CONDUCT.md --config '{\\"githubAccount\\":\\"mycompany\\",\\"copyrightHolder\\":\\"Great Company\\"}'",
      "npx baldrick-whisker@latest render --no-ext elm.json github:flarebyte:baldrick-reserve:template/elm/gitignore.hbs .gitignore.sh",
      "npx baldrick-whisker@latest render --no-ext elm.json github:flarebyte:baldrick-reserve:template/editorconfig.hbs .editorconfig.sh",
      "npx baldrick-whisker@latest render elm.json github:flarebyte:baldrick-reserve:template/elm/contributing.hbs CONTRIBUTING.md",
      "npx baldrick-whisker@latest render elm.json github:flarebyte:baldrick-reserve:template/elm/pull-request-template.hbs .github/pull_request_template.md",
      "npx baldrick-whisker@latest render github:flarebyte:baldrick-reserve:data/glossary.yaml github:flarebyte:baldrick-reserve:template/glossary.hbs GLOSSARY.md",
      "test -f \\"DECISIONS.md\\" || npx baldrick-whisker@latest render elm.json github:flarebyte:baldrick-reserve:template/decisions.hbs DECISIONS.md",
    ],
    "name": "whisker-norm",
    "parentMakeTask": undefined,
    "title": "Normalize the project with baldrick-whisker",
  },
  Object {
    "makeLines": Array [
      "npx baldrick-elm@latest generate -f lib -ga 'mycompany' -ch 'Great Company' -cy 2020 -l BSD3 && make md-fix",
    ],
    "name": "norm",
    "parentMakeTask": "whisker-norm",
    "title": "Normalize the code structure",
  },
]
`;
